File: examples/ciflypy-examples/ciflypy_examples/nearest_dsep.py
import ciflypy as cf
import ciflypy_examples.utils as utils

ruletables = utils.get_ruletable_path()
ancestors_table = cf.Ruletable(str(ruletables / "ancestors_admg.txt"))
closure_table = cf.Ruletable(str(ruletables / "closure_admg.txt"))

def find_nearest_separator(g, X, Y, I, R):
    A = cf.reach(g, {"X": X + Y + I}, ancestors_table)
    Z0 = set(R).intersection(set(A) - set(X + Y))
    Xstar = cf.reach(g, {"X": X, "Z": Z0, "A": A}, closure_table)
    if set(Xstar).intersection(Y):
        return None
    return list(Z0.intersection(Xstar).union(I))

File: examples/ciflyr-examples/R/nearestDsep.R
library("ciflyr")
library("here")

source(here("R", "utils.R"))

ruletables <- getRuletablePath()
ancestorsTable <- parseRuletable(file.path(ruletables, "ancestors_admg.txt"))
closureTable = parseRuletable(file.path(ruletables, "closure_admg.txt"))

findNearestSeparator <- function(g, X, Y, I, R) {
    A <- reach(g, list("X" = c(X, Y, I)), ancestorsTable)
    Z0 <- intersect(R, setdiff(A, c(X, Y)))
    Xstar <- reach(g, list("X" = X, "Z" = Z0, "A" = A), closureTable)
    if (length(intersect(Xstar, Y)) > 0) {
        return (NULL)
    }
    return (union(intersect(Z0, Xstar), I))
}
